File: example/display_robot_states.cpp
/**
 * @example display_robot_states.cpp
 * Print received robot states without enabling the robot.
 * @copyright Copyright (C) 2016-2021 Flexiv Ltd. All Rights Reserved.
 * @author Flexiv
 */

#include <flexiv/Robot.hpp>
#include <flexiv/Exception.hpp>
#include <flexiv/Log.hpp>
#include <flexiv/Utility.hpp>

#include <iostream>
#include <thread>

/** User-defined periodic task @ 1Hz */
void periodicTask(flexiv::Robot* robot, flexiv::Log* log)
{
    // Data struct for storing robot states
    flexiv::RobotStates robotStates;

    try {
        while (true) {
            // Wake up every second to do something
            std::this_thread::sleep_for(std::chrono::seconds(1));

            // Get robot states
            robot->getRobotStates(robotStates);

            // Print all robot states in JSON format using the built-in ostream
            // operator overloading
            log->info("");
            std::cout << robotStates << std::endl;
        }

    } catch (const flexiv::Exception& e) {
        log->error(e.what());
        return;
    }
}

void printHelp()
{
    // clang-format off
    std::cout << "Required arguments: [robot IP] [local IP]" << std::endl;
    std::cout << "    robot IP: address of the robot server" << std::endl;
    std::cout << "    local IP: address of this PC" << std::endl;
    std::cout << "Optional arguments: None" << std::endl;
    std::cout << std::endl;
    // clang-format on
}

int main(int argc, char* argv[])
{
    // Log object for printing message with timestamp and coloring
    flexiv::Log log;

    // Parse Parameters
    //=============================================================================
    if (argc < 3
        || flexiv::utility::programArgsExistAny(argc, argv, {"-h", "--help"})) {
        printHelp();
        return 1;
    }

    // IP of the robot server
    std::string robotIP = argv[1];

    // IP of the workstation PC running this program
    std::string localIP = argv[2];

    try {
        // RDK Initialization
        //=============================================================================
        // Instantiate robot interface
        flexiv::Robot robot(robotIP, localIP);

        // Clear fault on robot server if any
        if (robot.isFault()) {
            log.warn("Fault occurred on robot server, trying to clear ...");
            // Try to clear the fault
            robot.clearFault();
            std::this_thread::sleep_for(std::chrono::seconds(2));
            // Check again
            if (robot.isFault()) {
                log.error("Fault cannot be cleared, exiting ...");
                return 1;
            }
            log.info("Fault on robot server is cleared");
        }

        // Enable the robot, make sure the E-stop is released before enabling
        log.info("Enabling robot ...");
        robot.enable();

        // Wait for the robot to become operational
        int secondsWaited = 0;
        while (!robot.isOperational()) {
            std::this_thread::sleep_for(std::chrono::seconds(1));
            if (++secondsWaited == 10) {
                log.warn(
                    "Still waiting for robot to become operational, please "
                    "check that the robot 1) has no fault, 2) is booted "
                    "into Auto mode");
            }
        }
        log.info("Robot is now operational");

        // Periodic Tasks
        //=============================================================================
        // Use std::thread to do scheduling since this example is used for both
        // Linux and Windows, and the latter does not support flexiv::Scheduler
        std::thread lowPriorityThread(std::bind(periodicTask, &robot, &log));

        // Properly exit thread
        lowPriorityThread.join();

    } catch (const flexiv::Exception& e) {
        log.error(e.what());
        return 1;
    }

    return 0;
}

File: example_py/display_robot_states.py
#!/usr/bin/env python

"""display_robot_states.py

Print received robot states without enabling the robot.
"""

__copyright__ = "Copyright (C) 2016-2021 Flexiv Ltd. All Rights Reserved."
__author__ = "Flexiv"

import time
import argparse

# Import Flexiv RDK Python library
# fmt: off
import sys
sys.path.insert(0, "../lib_py")
import flexivrdk
# fmt: on


def main():
    # Parse Arguments
    # =============================================================================
    argparser = argparse.ArgumentParser()
    argparser.add_argument('robot_ip', help='IP address of the robot server')
    argparser.add_argument('local_ip', help='IP address of this PC')
    args = argparser.parse_args()

    # Define alias
    # =============================================================================
    robot_states = flexivrdk.RobotStates()
    log = flexivrdk.Log()
    mode = flexivrdk.Mode

    try:
        # RDK Initialization
        # =============================================================================
        # Instantiate robot interface
        robot = flexivrdk.Robot(args.robot_ip, args.local_ip)

        # Clear fault on robot server if any
        if robot.isFault():
            log.warn("Fault occurred on robot server, trying to clear ...")
            # Try to clear the fault
            robot.clearFault()
            time.sleep(2)
            # Check again
            if robot.isFault():
                log.error("Fault cannot be cleared, exiting ...")
                return
            log.info("Fault on robot server is cleared")

        # Enable the robot, make sure the E-stop is released before enabling
        log.info("Enabling robot ...")
        robot.enable()

        # Wait for the robot to become operational
        seconds_waited = 0
        while not robot.isOperational():
            time.sleep(1)
            seconds_waited += 1
            if seconds_waited == 10:
                log.warn(
                    "Still waiting for robot to become operational, please "
                    "check that the robot 1) has no fault, 2) is booted "
                    "into Auto mode")

        log.info("Robot is now operational")

        # Application-specific Code
        # =============================================================================
        while True:
            robot.getRobotStates(robot_states)
            log.info(" ")
            # Round all list values to 2 decimals
            # fmt: off
            print("q: ",  ['%.2f' % i for i in robot_states.q])
            print("theta: ", ['%.2f' % i for i in robot_states.theta])
            print("dq: ", ['%.2f' % i for i in robot_states.dq])
            print("dtheta: ", ['%.2f' % i for i in robot_states.dtheta])
            print("tau: ", ['%.2f' % i for i in robot_states.tau])
            print("tau_des: ", ['%.2f' % i for i in robot_states.tauDes])
            print("tau_dot: ", ['%.2f' % i for i in robot_states.tauDot])
            print("tau_ext: ", ['%.2f' % i for i in robot_states.tauExt])
            print("tcp_pose: ", ['%.2f' % i for i in robot_states.tcpPose])
            print("tcp_pose_d: ", ['%.2f' % i for i in robot_states.tcpPoseDes])
            print("tcp_velocity: ", ['%.2f' % i for i in robot_states.tcpVel])
            print("camera_pose: ", ['%.2f' % i for i in robot_states.camPose])
            print("flange_pose: ", ['%.2f' % i for i in robot_states.flangePose])
            print("end_link_pose: ", ['%.2f' % i for i in robot_states.endLinkPose])
            print("F_ext_tcp_frame: ", ['%.2f' % i for i in robot_states.extForceInTcpFrame])
            print("F_ext_base_frame: ", ['%.2f' % i for i in robot_states.extForceInBaseFrame])
            time.sleep(1)
            # fmt: on

    except Exception as e:
        # Print exception error message
        log.error(str(e))


if __name__ == "__main__":
    main()
